cpp/include/cuml/experimental/fil/detail/gpu_introspection.hpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cuda_runtime_api.h>
+#include <cuml/experimental/fil/detail/index_type.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/cuda_check.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/device_id.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/device_type.hpp>
+#include <vector>
+
+namespace ML {
+namespace experimental {
+namespace fil {
+namespace detail {
+
+inline auto get_max_shared_mem_per_block(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto static cache = std::vector<int>{};
+  if (cache.size() == 0) {
+    auto device_count = int{};
+    raft_proto::cuda_check(cudaGetDeviceCount(&device_count));
+    cache.resize(device_count);
+    for (auto dev=0; dev < device_count; ++dev) {
+      raft_proto::cuda_check(
+        cudaDeviceGetAttribute(
+          &(cache[dev]),
+          cudaDevAttrMaxSharedMemoryPerBlockOptin,
+          dev
+        )
+      );
+    }
+  }
+  return index_type(cache.at(device_id.value()));
+}
+
+inline auto get_sm_count(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto static cache = std::vector<int>{};
+  if (cache.size() == 0) {
+    auto device_count = int{};
+    raft_proto::cuda_check(cudaGetDeviceCount(&device_count));
+    cache.resize(device_count);
+    for (auto dev=0; dev < device_count; ++dev) {
+      raft_proto::cuda_check(
+        cudaDeviceGetAttribute(
+          &(cache[dev]),
+          cudaDevAttrMultiProcessorCount,
+          dev
+        )
+      );
+    }
+  }
+  return index_type(cache.at(device_id.value()));
+}
+
+inline auto get_max_threads_per_block(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto result = int{};
+  raft_proto::cuda_check(
+    cudaDeviceGetAttribute(
+      &result,
+      cudaDevAttrMaxThreadsPerBlock,
+      device_id.value()
+    )
+  );
+  return index_type(result);
+}
+
+inline auto get_max_threads_per_sm(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto result = int{};
+  raft_proto::cuda_check(
+    cudaDeviceGetAttribute(
+      &result,
+      cudaDevAttrMaxThreadsPerMultiProcessor,
+      device_id.value()
+    )
+  );
+  return index_type(result);
+}
+
+inline auto get_max_shared_mem_per_sm(raft_proto::device_id<raft_proto::device_type::gpu> device_id) {
+  auto static cache = std::vector<int>{};
```

Comment:
Is the introspection function called by only one thread? If multiple threads might call the function, we should either:
1) Maintain a thread-local cache OR
2) Use `std::mutex` to protect concurrent access